pkg/vault/contracts/CompositeLiquidityRouter.sol

```diff
@@ -482,10 +484,50 @@ contract CompositeLiquidityRouter is ICompositeLiquidityRouter, BatchRouterCommo
         // Loads a Set with all amounts to be inserted in the nested pools, so we don't need to iterate in the tokens
         // array to find the child pool amounts to insert.
         for (uint256 i = 0; i < tokensIn.length; ++i) {
+            if (params.maxAmountsIn[i] == 0) {
+                continue;
+            }
+
             _currentSwapTokenInAmounts().tSet(tokensIn[i], params.maxAmountsIn[i]);
+            _currentSwapTokensIn().add(tokensIn[i]);
         }
 
-        IERC20[] memory parentPoolTokens = _vault.getPoolTokens(params.pool);
+        (uint256[] memory amountsIn, ) = _addLiquidityRecursive(params.pool, params);
```

Comment:
At first glance it looks like you could just have `_addLiquidityRecursive` take a single parameter, since the pool is already in the params - but you can't, because sometimes it's the child pool, different from the one in params. Would be good to document that somewhere (maybe where the function's defined).

Reply:
We might also note that the Vault explicitly prohibits "Linear pool-style" recursion, where a pool contains its own BPT, so that case is impossible and we don't need to handle it.